source/exe/main.cc

```diff
@@ -24,32 +9,22 @@
  * deployment such as initializing signal handling. It calls main_common
  * after setting up command line options.
  */
-int main(int argc, char** argv) {
-#ifdef ENVOY_HANDLE_SIGNALS
-  // Enabled by default. Control with "bazel --define=signal_trace=disabled"
-  Envoy::SignalAction handle_sigs;
-#endif
-
+int main(int argc, const char** argv) {
+  try {
 #ifdef ENVOY_HOT_RESTART
-  // Enabled by default, except on OS X. Control with "bazel --define=hot_restart=disabled"
-  const Envoy::OptionsImpl::HotRestartVersionCb hot_restart_version_cb =
-      [](uint64_t max_num_stats, uint64_t max_stat_name_len) {
-        return Envoy::Server::HotRestartImpl::hotRestartVersion(max_num_stats, max_stat_name_len);
-      };
+    Envoy::MainCommon main_common(argc, argv, true);
 #else
-  const Envoy::OptionsImpl::HotRestartVersionCb hot_restart_version_cb = [](uint64_t, uint64_t) {
-    return "disabled";
-  };
+    Envoy::MainCommon main_common(argc, argv, false);
 #endif
-
-  std::unique_ptr<Envoy::OptionsImpl> options;
-  try {
-    options = std::make_unique<Envoy::OptionsImpl>(argc, argv, hot_restart_version_cb,
-                                                   spdlog::level::info);
+    return main_common.run() ? 0 : 1;
   } catch (const Envoy::NoServingException& e) {
-    return 0;
+    return EXIT_SUCCESS;
   } catch (const Envoy::MalformedArgvException& e) {
-    return 1;
+    std::cerr << "MalformedArgvException: " << e.what() << std::endl;
```

Comment:
What's the reasoning behind these new catch blocks? Where these things caught elsewhere and now are caught here? My main comment is that the errors we print should be user friendly so I don't think we should print the name of the exception which the user doesn't care about. Perhaps just "Envoy startup failure: ..."?

Reply:
This isn't extra; it's just moved.  It was previously on main_common.cc:92.  I removed the extra printing though, and handles the case in Server::InstanceImpl::InstanceImpl where the config fails to parse, e.g. because it is empty, which is tested in hotrestart_integration_test.  I guess an invalid json config could also cause this, or failing to open the log file.

I think it makes sense to handle EnvoyException in main.cc.  Or, we could re-cast the exception in MainCommon's ctor as one that was already handled in main.cc.  WDYT?

Reply:
I think it's fine either way. The main thing I would watch out for is that we don't catch exceptions when the run loop starts running. I realize that wasn't really clearly codified, and feel free to add more comments, but the idea is that we can throw and exit(1) during startup, but once the main loop starts up any exceptions which aren't explicitly handled should crash the process w/ core dump. That's the way the code works today so I would just be careful that we maintain that behavior.